homeassistant/components/simplisafe/camera.py

```diff
@@ -0,0 +1,243 @@
+    ) -> bytes | None:
+        """Return bytes of camera image."""
+        if self._attr_cached_image is not None:
+            return self._attr_cached_image
+
+        if self._attr_image_url is None:
+            return None
+
+        self._attr_cached_image = await self._simplisafe.async_media_request(
+            self._attr_image_url.replace("{&width}", "&width=" + str(width))
+        )
+        return self._attr_cached_image
+
+    def _register_services(self) -> None:
+        """Register the Outdoor Camera hass service calls."""
+
+        def _write_image(to_file: str, image_data: bytes) -> None:
+            """Executor helper to write image."""
+            os.makedirs(os.path.dirname(to_file), exist_ok=True)
+            with open(to_file, "wb") as img_file:
+                img_file.write(image_data)
+
+        async def save_image_handler(call: ServiceCall) -> None:
+            if self._attr_image_url is None:
+                return
+
+            width = call.data.get(ATTR_WIDTH)
+            filename: Template = call.data.get(ATTR_FILENAME)  # type: ignore[assignment]
+            filename.hass = self._hass
+            snapshot_file = filename.async_render(variables={ATTR_ENTITY_ID: self})
+
+            try:
+                snapshot = await self._simplisafe.async_media_request(
+                    self._attr_image_url.replace("{&width}", "&width=" + str(width))
+                )
+            except SimplipyError as err:
+                raise HomeAssistantError(
+                    f'Error fetching motion media "{self._system.system_id}": {err}'
+                ) from err
+
+            try:
+                await self._hass.async_add_executor_job(
+                    _write_image, snapshot_file, snapshot
+                )
+            except OSError as err:
+                LOGGER.error("Can't write image to file: %s", err)
+
+        async def save_clip_handler(call: ServiceCall) -> None:
+            if self._attr_clip_url is None:
+                return
+
+            filename: Template = call.data.get(ATTR_FILENAME)  # type: ignore[assignment]
+            filename.hass = self._hass
+            clip_file = filename.async_render(variables={ATTR_ENTITY_ID: self})
+
+            try:
+                clip = await self._simplisafe.async_media_request(self._attr_clip_url)
+            except SimplipyError as err:
+                raise HomeAssistantError(
+                    f'Error fetching motion media "{self._system.system_id}": {err}'
+                ) from err
+
+            try:
+                await self._hass.async_add_executor_job(_write_image, clip_file, clip)
+            except OSError as err:
+                LOGGER.error("Can't write image to file: %s", err)
+
+        self._hass.services.async_register(
+            DOMAIN,
+            SERVICE_OC_IMAGE,
+            save_image_handler,
+            schema=SERVICE_OC_IMAGE_SCHEMA,
+        )
+
+        self._hass.services.async_register(
```

Comment:
Loop through these service registration calls (since they're almost the same).

Reply:
Only DOMAIN is common.  The other 3 (of 4) are unique.  Seems to me the loop would be complicated.